### ISSUE TITLE
[provenance] Allow users to create RO-Crate archives without archiving experiments

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1412]
At the moment users need to archive experiments to get an RO-Crate archive. But users shouldn't need to archive an experiment to get a provenance archive. We should enable something like `autosubmit run --provenance`, and also another option like `autosubmit provenance --etc..`, so that users can run and get a provenance archive at the end, or get the archive for the latest run without archiving the experiment.

The RO-Crate file will probably need some naming schema to differentiate each file (e.g. `rocrate-$date?.zip` or `rocrate-$runid?.zip`, etc.). This needs to be in the changelog, have basic tests, and re-use the existing functions for RO-Crate in Autosubmit. There's also a couple of places where RO-Crate is mentioned in our docs, we can update where the RO-Crate functionality is described (I think the other page is general about provenance, not sure if we need to touch that one).

A final test to review the change would be to upload it to WorkflowHub, or use [`runcrate`](https://github.com/ResearchObject/runcrate) or the new [validation tool they are working on](https://github.com/crs4/rocrate-validator/).

I've assigned @edgano but I guess work might be done by him or @apuiggro. Assigning milestone 4.2.0 just so we can discuss about it, but I believe it will be for 4.3.0 or later.

**Comment by kinow** [comment]
In GitLab by @apuiggro on Oct 10, 2024, 09:38

Hi @kinow, I finally learned how to properly debugg the code and now I'm testing modifications made. Currently I'm having issues identifying why the code stops after parse_args() function. 

I'm using this Debugger conf for VsCode: 

`{
    "version": "0.2.0",
    "configurations": [

        {
            "name": "python : autosubmit",
            "type": "python",
            "request": "launch",
            "program": "/home/albertpuiggros/autosubmit/venv/bin/autosubmit",
            "args": ["provenance", "--rocrate", "a000"], 
            "console": "integratedTerminal"
        }
    ]
}`

It stops at:

![Captura_de_pantalla_2024-10-10_093416](https://earth.bsc.es/gitlab/es/autosubmit/uploads/b5669ca6ec0a01bbb90de486a29fe3d4/Captura_de_pantalla_2024-10-10_093416.png)

All variables seems to be correclty stated in the praser function: 

![Captura_de_pantalla_2024-10-10_093539](https://earth.bsc.es/gitlab/es/autosubmit/uploads/5f843a950490b2b8ca889430119c4361/Captura_de_pantalla_2024-10-10_093539.png)

The next "Step over" sends me to main() and stops at the selected (yellow) line: 

![Captura_de_pantalla_2024-10-10_093713](https://earth.bsc.es/gitlab/es/autosubmit/uploads/66ade2d26e7341f81c5a657b2abfe4c7/Captura_de_pantalla_2024-10-10_093713.png)

I think maybe I'm debugging the code worng as no error pops up. 
Working on it now...

**Comment by kinow** [comment]
In GitLab by @apuiggro on Oct 14, 2024, 11:53

I have completed the draft implementation of the provenance function without the need for archiving the experiment. The code is still preliminary and needs to be reviewed. I also created a unit test to check the behavior of the function, and everything seems to be functioning well so far. One thing we still need to discuss is where to store the RO-Crate file within the experiment folder; currently, it is located in the ASLOGS folder.

Function "provenance()": 
![Captura_de_pantalla_2024-10-14_111703](https://earth.bsc.es/gitlab/es/autosubmit/uploads/4a840eac1c7422887db08cf6b227f1e2/Captura_de_pantalla_2024-10-14_111703.png)

The test: 

![Captura_de_pantalla_2024-10-14_115305](https://earth.bsc.es/gitlab/es/autosubmit/uploads/a63efa0314882d92cdb56bf9a409dae5/Captura_de_pantalla_2024-10-14_115305.png)

![Captura_de_pantalla_2024-10-14_115316](https://earth.bsc.es/gitlab/es/autosubmit/uploads/947d383ca74ae5d1723ab89c4474d610/Captura_de_pantalla_2024-10-14_115316.png)

![Captura_de_pantalla_2024-10-14_115321](https://earth.bsc.es/gitlab/es/autosubmit/uploads/85948151720f9238a300358fb0db6ce9/Captura_de_pantalla_2024-10-14_115321.png)

![Captura_de_pantalla_2024-10-14_115326](https://earth.bsc.es/gitlab/es/autosubmit/uploads/c0822629c4dac9af81ec879c851832ce/Captura_de_pantalla_2024-10-14_115326.png)

**Comment by kinow** [comment]
Just updating the ticket with what we discussed today in person at the office.

- We configured an experiment with RO-Crate successfully :tada: , it's on Albert's laptop, `a000`
- The code worked flawlessly :clap: 

>Currently, the rocrate file is defined as `{expid}-{date}.zip`. Should I modify it and, for example, add the runid to the existing structure (`{expid}-{date}-{runid}.zip`) ?

As discussed, let's

- Add runID inside the RO-Crate file somehow. It can be a separate issue if you think it'd be better
- Maybe go with `{expid}-{date}`.zip` but with date/timestamp going down to milliseconds, related to #1368

**Comment by kinow** [comment]
In GitLab by @apuiggro on Oct 28, 2024, 10:01

For now, I modified the `date` to include hours, minutes, and seconds, ensuring that each generated name will be distinguishable. I still don’t know exactly how to proceed with the runid—I’m not sure where to retrieve it from or where it is defined. Can't find any documentation about it. 

Also, I've successfully generated provenance for SUNSET when executing multiple recipes in parallel with Autosubmit. I wanted to simultaneously test provenance for both SUNSET and Autosubmit, but the current version of the Autosubmit machines (bscesautosubmit01 / bsceautosubmit01) doesn’t support generating RO-Crate objects. Should I open an issue for this with @kinow?

**Comment by kinow** [comment]
>For now, I modified the `date` to include hours, minutes, and seconds, ensuring that each generated name will be distinguishable. I still don’t know exactly how to proceed with the runid—I’m not sure where to retrieve it from or where it is defined. Can't find any documentation about it.

Great. I think we still don't have a draft merge request? We need one now to start reviewing the code and commenting on it. We can use that MR to discuss the runid.

> Also, I've successfully generated provenance for SUNSET when executing multiple recipes in parallel with Autosubmit. I wanted to simultaneously test provenance for both SUNSET and Autosubmit, but the current version of the Autosubmit machines (bscesautosubmit01 / bsceautosubmit01) doesn’t support generating RO-Crate objects. Should I open an issue for this with @kinow?

Hmmm, I see the issue. RO-Crate was added in 4.0.100 (released in 2023)) but the bscesautosubmit0N machines have only up to 4.0.98.

@apuiggro couldn't you use bsceshub0N machines instead? I believe the plan is to decommission bscesautosubmit machines and replace by the bsceshubs.

**Comment by kinow** [comment]
In GitLab by @apuiggro on Oct 28, 2024, 12:15

Okey! I will create a draft merge request!
